### PR TITLE
Enable `--warn-redundant-casts` option in mypy

### PR DESCRIPTION
**Pull request by DMRobertson**
Doesn't do much but helps me sleep better at night.

### CODE REVIEW

File: synapse/storage/engines/postgres.py [hunk 1]
```diff
@@ -77,7 +77,7 @@ def check_database(
         # docs: The number is formed by converting the major, minor, and
         # revision numbers into two-decimal-digit numbers and appending them
         # together. For example, version 8.1.5 will be returned as 80105
-        self._version = cast(int, db_conn.server_version)
```
Comment (DMRobertson) [inline synapse/storage/engines/postgres.py:80]: https://github.com/python/typeshed/pull/7834 is the underlying typeshed change

**Review by clokep — COMMENTED** [review]
Looks reasonable, but CI is failing on these lines? Do we need to bump mypy somewhere?

**Comment by DMRobertson** [comment]
> Do we need to bump mypy somewhere?

I think I just rushed the type-ignore without checking it worked. Not sure why I didn't see the `Version` one locally though. Maybe some kind of caching? :shrug:

File: scripts-dev/release.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def _prepare() -> None:
     # Switch to the release branch.
     # Cast safety: parse() won't return a version.LegacyVersion from our
     # version string format.
```
Comment (clokep) [inline scripts-dev/release.py:178]: I think we can remove this comment!

```suggestion
```

File: synapse/storage/database.py [hunk 3]
```diff
@@ -667,7 +667,8 @@ def new_transaction(
                 )
         # also check variables referenced in func's closure
         if inspect.isfunction(func):
-            f = cast(types.FunctionType, func)
+            # Keep the cast for now---it helps PyCharm to understand what `func` is.
```
Comment (clokep) [inline synapse/storage/database.py:670]: I guess we would need this until PyCharm had better understanding of `inspect.isfunction` or something?

**Review by clokep — APPROVED** [review]

File: synapse/storage/database.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/storage/database.py:670]: Yeah. There's a generic mechanism for this kind of pattern (TypeGuard). I'm not sure how well IntelliJ understands it though.

**Pull request merged**